caas/kubernetes/provider/cloud.go

```diff
@@ -294,9 +190,9 @@ func (p kubernetesEnvironProvider) FinalizeCloud(ctx environs.FinalizeCloudConte
 		},
 	}
 
-	_, err = UpdateKubeCloudWithStorage(&cld, storageUpdateParams)
+	cld, err = UpdateKubeCloudWithStorage(cld, storageUpdateParams)
```

Comment:
Why change the param from a pointer?
Also, on error, it's more idiomatic to return the zero value for the result param.

Reply:
So that side effects are obvious and we aren't just altering the parameters passed in. We also establish a good delta now.

Returning what we have built up in the cloud instead of a non zero now because we have done processing up until that point and there is still a valid struct. We also don't start out with an invalid struct so there is no harm here.

I wouldn't call returning the nil value "idiomatic". The more idiomatic solution is that if your function is completely hosed then the safe thing is to return nil values with the error. But in this case we are starting with a known working value and just updating properties. There are plenty of examples in the std library of this behavior.

Reply:
So the problem with blindly returning partially formed values is that it is implementation specific and to the consumer of the function, undefined behaviour. If you make it part of the contract of the function, to say, on error, this function can return partial output in this form, then that is ok.

Reply:
Additionally, if you return partial output from a function, that partial output needs to be unit tested. So by doing so, you essentially open yourself up to testing a lot more.